JSON file generation failed while deploying the a pipeline

##### Issue Type

<!--- Please specify the issue type to help us categorize the issue, mention any one of the below types -->

- Bug Report


##### Summary

<!--- Provide a brief summary of the issue -->

While triggering a pipeline that contains multiple stages with automatic approval for the next stage, we are seeing an error in generating the JSON file.
The pipeline is successful, but our internal deployment tool can't receive the JSON file because of the generation failure, the previous stages and different pipelines without the multiple automatic approval works without issues.

##### Environment

<!--- 
Describe the environment in which you encountered this issue.
Details of the environment GoCD is running in can be fetched from the GoCD support api - http://<go-server-host>/go/api/support.
If you are pasting plain text, please surround it with 3 backticks on each side

```
```
-->

###### Basic environment details

<!--- We recommend providing the for us to reproduce the issue quicker -->

* Go Version: `23.5.0 (18179-7702b283accd1f90f014f0087aa2e9bd8baf4a97)`
* JAVA Version: `17.0.9`
* OS: `Linux 4.19.0-23-cloud-amd64`


###### Additional Environment Details

<!-- More environment details captured from the support API or other sources can be shared here -->

##### Expected Results
<!--- Tell us what should happen -->

We expected it to behave normally like the previous stage:

```
2024-04-05 13:33:49,061 INFO  [qtp380095890-28] ScheduleService:226 - [Stage Schedule] Scheduling stage sandbox-europe-west4 for pipeline ppro-girogate
2024-04-05 13:33:50,786 INFO  [108@MessageListener for WorkFinder] BuildAssignmentService:190 - [Agent Assignment] Assigned job [JobIdentifier[ppro-girogate, 7, 1.7.0, sandbox-europe-west4, 1, deploy-gcp, 51499]] to agent [Agent [sandbox-europe-west4-gocd-agent-0, 172.20.0.29, 7a8015ac-7c44-4230-98e0-78fc991535e3]]
```

##### Actual Results
<!--- Tell us what happens instead -->

See the log snippet

##### Possible Fix
<!--- Not obligatory, but suggest a fix or reason for the bug -->

##### Log snippets
<!--- 
If you could find any information/exceptions from the logs do provide it here. 
Do mask any confidential information which you don't want to be shared.
Remember to surround them in 3 backticks like this —
-->

```
2024-04-05 13:36:41,126 INFO  [qtp380095890-26] ScheduleService:226 - [Stage Schedule] Scheduling stage opglive-europe-west1 for pipeline ppro-girogate
2024-04-05 13:36:41,848 ERROR [qtp380095890-38] JsonOutputWriter:85 - There was an error generating JSON
java.lang.RuntimeException: java.lang.NullPointerException: Cannot invoke "java.sql.Timestamp.getTime()" because the return value of "com.thoughtworks.go.domain.Stage.getLastTransitionedTime()" is null
        at com.thoughtworks.go.api.base.JsonOutputWriter$JsonOutputWriterUsingJackson$ConsumerWhichThrows.accept(JsonOutputWriter.java:291)
        at com.thoughtworks.go.api.base.JsonOutputWriter$JsonOutputWriterUsingJackson.withExceptionHandling(JsonOutputWriter.java:278)
        at com.thoughtworks.go.api.base.JsonOutputWriter$JsonOutputWriterUsingJackson.forTopLevelObject(JsonOutputWriter.java:298)
        at com.thoughtworks.go.api.base.JsonOutputWriter.lambda$forTopLevelObject$0(JsonOutputWriter.java:59)
        at com.thoughtworks.go.api.base.JsonOutputWriter.bufferWriterAndFlushWhenDone(JsonOutputWriter.java:80)
        at com.thoughtworks.go.api.base.JsonOutputWriter.forTopLevelObject(JsonOutputWriter.java:57)
        at com.thoughtworks.go.api.ControllerMethods.writerForTopLevelObject(ControllerMethods.java:91)
        at com.thoughtworks.go.apiv3.stageinstance.StageInstanceControllerV3.instanceByCounter(StageInstanceControllerV3.java:160)
        at spark.RouteImpl$1.handle(RouteImpl.java:72)
        at spark.RouteImpl$1.handle(RouteImpl.java:72)
        at spark.http.matching.Routes.execute(Routes.java:61)
        at spark.http.matching.MatcherFilter.doFilter(MatcherFilter.java:130)
        at spark.servlet.SparkFilter.doFilter(SparkFilter.java:173)
        at com.thoughtworks.go.spark.SparkPreFilter.doFilter(SparkPreFilter.java:62)
        at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:202)
        at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1635)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:527)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:131)
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:616)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:223)
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1580)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:131)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
        at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:680)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:223)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1384)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:176)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:484)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1553)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:174)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1306)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:129)
        at org.eclipse.jetty.server.Dispatcher.forward(Dispatcher.java:215)
        at org.eclipse.jetty.server.Dispatcher.forward(Dispatcher.java:135)
        at org.tuckey.web.filters.urlrewrite.NormalRewrittenUrl.doRewrite(NormalRewrittenUrl.java:213)
        at org.tuckey.web.filters.urlrewrite.RuleChain.handleRewrite(RuleChain.java:171)
        at org.tuckey.web.filters.urlrewrite.RuleChain.doRules(RuleChain.java:145)
        at org.tuckey.web.filters.urlrewrite.UrlRewriter.processRequest(UrlRewriter.java:92)
        at org.tuckey.web.filters.urlrewrite.UrlRewriteFilter.doFilter(UrlRewriteFilter.java:381)
        at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:202)
        at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1635)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317)
        at com.thoughtworks.go.server.web.FlashLoadingFilter.doFilterInternal(FlashLoadingFilter.java:39)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:208)
        at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:185)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317)
        at com.thoughtworks.go.server.newsecurity.filters.DenyIfRefererIsNotFilesFilter.doFilterInternal(DenyIfRefererIsNotFilesFilter.java:52)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:214)
        at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:185)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317)
        at com.thoughtworks.go.server.newsecurity.filters.VerifyAuthorityFilter.doFilterInternal(VerifyAuthorityFilter.java:57)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)

```

##### Code snippets/Screenshots
<!--- 
If you would like to share any code/screenshots please add it here.
Do mask any confidential information which you don't want to be shared.
Remember to surround them in 3 backticks like this —

```
long log lines go here
```
-->

![image](https://github.com/gocd/gocd/assets/97034554/500466d2-c87d-478f-8dba-34f3d6fd52ff)


##### Any other info
<!--- Provide any other information you would like to share to explain this issue more better -->

- It sounds to me like the GoCD server version change is probably a coincidence, however which GoCD server version were you running previously that you hadn't seen this issue with?

 R = We had version 19.12 before, and we performed a upgrade to 23.5
 
- OK, this is interesting, thanks. You might want to check the logs of the gocd-server during the period that opglive-europe-west1 was running/completing to see if anything unusual happened, like the server restarting; or errors/stack traces in the logs that might be relevant. Otherwise sharing the logs during the period that pipeline was running would be useful (may need to check and partially redact them).

Those are the logs during that run, there are a few hanging jobs that are not related to this one (but they were also successfully completed and are hanging in the UI after the server upgrade):

```
2024-04-05 13:36:41,126 INFO  [qtp380095890-26] ScheduleService:226 - [Stage Schedule] Scheduling stage opglive-europe-west1 for pipeline ppro-girogate
2024-04-05 13:36:41,848 ERROR [qtp380095890-38] JsonOutputWriter:85 - There was an error generating JSON
java.lang.RuntimeException: java.lang.NullPointerException: Cannot invoke "java.sql.Timestamp.getTime()" because the return value of "com.thoughtworks.go.domain.Stage.getLastTransitionedTime()" is null
        at com.thoughtworks.go.api.base.JsonOutputWriter$JsonOutputWriterUsingJackson$ConsumerWhichThrows.accept(JsonOutputWriter.java:291)
        at com.thoughtworks.go.api.base.JsonOutputWriter$JsonOutputWriterUsingJackson.withExceptionHandling(JsonOutputWriter.java:278)
        at com.thoughtworks.go.api.base.JsonOutputWriter$JsonOutputWriterUsingJackson.forTopLevelObject(JsonOutputWriter.java:298)
        at com.thoughtworks.go.api.base.JsonOutputWriter.lambda$forTopLevelObject$0(JsonOutputWriter.java:59)
        at com.thoughtworks.go.api.base.JsonOutputWriter.bufferWriterAndFlushWhenDone(JsonOutputWriter.java:80)
        at com.thoughtworks.go.api.base.JsonOutputWriter.forTopLevelObject(JsonOutputWriter.java:57)
        at com.thoughtworks.go.api.ControllerMethods.writerForTopLevelObject(ControllerMethods.java:91)
        at com.thoughtworks.go.apiv3.stageinstance.StageInstanceControllerV3.instanceByCounter(StageInstanceControllerV3.java:160)
        at spark.RouteImpl$1.handle(RouteImpl.java:72)
        at spark.RouteImpl$1.handle(RouteImpl.java:72)
        at spark.http.matching.Routes.execute(Routes.java:61)
        at spark.http.matching.MatcherFilter.doFilter(MatcherFilter.java:130)
        at spark.servlet.SparkFilter.doFilter(SparkFilter.java:173)
        at com.thoughtworks.go.spark.SparkPreFilter.doFilter(SparkPreFilter.java:62)
        at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:202)
        at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1635)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:527)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:131)
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:616)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:223)
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1580)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:131)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
        at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:680)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:223)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1384)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:176)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:484)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1553)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:174)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1306)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:129)
        at org.eclipse.jetty.server.Dispatcher.forward(Dispatcher.java:215)
        at org.eclipse.jetty.server.Dispatcher.forward(Dispatcher.java:135)
        at org.tuckey.web.filters.urlrewrite.NormalRewrittenUrl.doRewrite(NormalRewrittenUrl.java:213)
        at org.tuckey.web.filters.urlrewrite.RuleChain.handleRewrite(RuleChain.java:171)
        at org.tuckey.web.filters.urlrewrite.RuleChain.doRules(RuleChain.java:145)
        at org.tuckey.web.filters.urlrewrite.UrlRewriter.processRequest(UrlRewriter.java:92)
        at org.tuckey.web.filters.urlrewrite.UrlRewriteFilter.doFilter(UrlRewriteFilter.java:381)
        at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:202)
        at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1635)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317)
        at com.thoughtworks.go.server.web.FlashLoadingFilter.doFilterInternal(FlashLoadingFilter.java:39)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:208)
        at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:185)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317)
        at com.thoughtworks.go.server.newsecurity.filters.DenyIfRefererIsNotFilesFilter.doFilterInternal(DenyIfRefererIsNotFilesFilter.java:52)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:214)
        at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:185)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317)
        at com.thoughtworks.go.server.newsecurity.filters.VerifyAuthorityFilter.doFilterInternal(VerifyAuthorityFilter.java:57)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:214)
        at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:185)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317)
        at com.thoughtworks.go.server.newsecurity.filters.AbstractUserEnabledCheckFilter.doFilterInternal(AbstractUserEnabledCheckFilter.java:66)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:214)
        at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:185)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at com.thoughtworks.go.server.newsecurity.filters.ThreadLocalUserFilter.doFilterInternal(ThreadLocalUserFilter.java:41)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317)
        at com.thoughtworks.go.server.newsecurity.filters.AccessTokenAuthenticationFilter.filterWhenSecurityEnabled(AccessTokenAuthenticationFilter.java:128)
        at com.thoughtworks.go.server.newsecurity.filters.AccessTokenAuthenticationFilter.doFilterInternal(AccessTokenAuthenticationFilter.java:94)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at com.thoughtworks.go.server.newsecurity.filters.AbstractBasicAuthenticationFilter.doFilterInternal(AbstractBasicAuthenticationFilter.java:56)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at com.thoughtworks.go.server.newsecurity.filters.AssumeAnonymousUserFilter.doFilterInternal(AssumeAnonymousUserFilter.java:64)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at com.thoughtworks.go.server.newsecurity.filters.AbstractReAuthenticationFilter.doFilterInternal(AbstractReAuthenticationFilter.java:74)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at com.thoughtworks.go.server.newsecurity.filters.InvalidateAuthenticationOnSecurityConfigChangeFilter.doFilterInternal(InvalidateAuthenticationOnSecurityConfigChangeFilter.java:105)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:214)
        at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:185)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:208)
        at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:185)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317)
        at com.thoughtworks.go.server.newsecurity.filters.AlwaysCreateSessionFilter.doFilterInternal(AlwaysCreateSessionFilter.java:40)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at com.thoughtworks.go.server.newsecurity.filters.AbstractSessionReduceIdleTimeoutFilter.doFilterInternal(AbstractSessionReduceIdleTimeoutFilter.java:44)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:214)
        at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:185)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at com.thoughtworks.go.server.newsecurity.filters.ModeAwareFilter.doFilter(ModeAwareFilter.java:76)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at com.thoughtworks.go.server.web.BackupFilter.doFilterInternal(BackupFilter.java:79)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
        at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:214)
        at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:177)
        at com.thoughtworks.go.server.newsecurity.filterchains.MainFilterChain.doFilter(MainFilterChain.java:79)
        at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:347)
        at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:263)
        at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:202)
        at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1635)
        at com.thoughtworks.go.server.web.DefaultHeadersFilter.doFilter(DefaultHeadersFilter.java:51)
        at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:210)
        at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1635)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:527)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:131)
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:598)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:223)
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1580)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:131)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
        at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:822)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:223)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1384)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:176)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:484)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1553)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:174)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1306)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:129)
        at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:192)
        at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:141)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
        at org.eclipse.jetty.server.Server.handle(Server.java:563)
        at org.eclipse.jetty.server.HttpChannel$RequestDispatchable.dispatch(HttpChannel.java:1598)
        at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:753)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:501)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:287)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:314)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
        at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
        at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:421)
        at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:390)
        at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:277)
        at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.run(AdaptiveExecutionStrategy.java:199)
        at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:411)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:969)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1194)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1149)
        at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.NullPointerException: Cannot invoke "java.sql.Timestamp.getTime()" because the return value of "com.thoughtworks.go.domain.Stage.getLastTransitionedTime()" is null
        at com.thoughtworks.go.apiv3.stageinstance.representers.StageRepresenter.toJSON(StageRepresenter.java:30)
        at com.thoughtworks.go.apiv3.stageinstance.StageInstanceControllerV3.lambda$instanceByCounter$6(StageInstanceControllerV3.java:160)
        at com.thoughtworks.go.api.base.JsonOutputWriter$JsonOutputWriterUsingJackson.lambda$forTopLevelObject$21(JsonOutputWriter.java:300)
        at com.thoughtworks.go.api.base.JsonOutputWriter$JsonOutputWriterUsingJackson$ConsumerWhichThrows.accept(JsonOutputWriter.java:289)
        ... 159 common frames omitted
2024-04-05 13:36:46,064 INFO  [108@MessageListener for WorkFinder] BuildAssignmentService:190 - [Agent Assignment] Assigned job [JobIdentifier[ppro-girogate, 7, 1.7.0, opglive-europe-west1, 1, deploy-gcp, 51500]] to agent [Agent [opglive-europe-west1-gocd-agent-0, 10.100.36.3, 781d843e-1f98-4142-ac1e-80b2f6a94cf5]]
2024-04-05 13:36:52,752 INFO  [goThreadPoolTaskScheduler-2] ConsoleActivityMonitor:111 - Job 'tech-docs/131/opglive-europe-west4/1/deploy-docker' hung for more than '2856' minutes
2024-04-05 13:36:52,753 INFO  [goThreadPoolTaskScheduler-2] ConsoleActivityMonitor:111 - Job 'grafana-dashboards/382/sandbox-europe-west4/1/deploy-docker' hung for more than '1505' minutes
2024-04-05 13:36:52,753 INFO  [goThreadPoolTaskScheduler-2] ConsoleActivityMonitor:111 - Job 'grafana-dashboards/382/opglive-europe-west4/1/deploy-docker' hung for more than '1504' minutes
2024-04-05 13:36:52,753 INFO  [goThreadPoolTaskScheduler-2] ConsoleActivityMonitor:111 - Job 'grafana-dashboards/382/stage/1/deploy-docker' hung for more than '1506' minutes
2024-04-05 13:36:52,753 INFO  [goThreadPoolTaskScheduler-2] ConsoleActivityMonitor:111 - Job 'grafana-dashboards/382/stage/1/deploy' hung for more than '1506' minutes
2024-04-05 13:37:03,479 INFO  [qtp380095890-31] BuildRepositoryRemoteImpl:96 - [Agent [opglive-europe-west1-gocd-agent-0, 10.100.36.3, 781d843e-1f98-4142-ac1e-80b2f6a94cf5, b78bd3e3-bc67-4df6-a3e5-b1d740507385]] is reporting status [Preparing] for [Build [ppro-girogate/7/opglive-europe-west1/1/deploy-gcp/51500]]
2024-04-05 13:37:09,135 INFO  [qtp380095890-31] BuildRepositoryRemoteImpl:96 - [Agent [opglive-europe-west1-gocd-agent-0, 10.100.36.3, 781d843e-1f98-4142-ac1e-80b2f6a94cf5, b78bd3e3-bc67-4df6-a3e5-b1d740507385]] is reporting status [Building] for [Build [ppro-girogate/7/opglive-europe-west1/1/deploy-gcp/51500]]
2024-04-05 13:37:22,753 INFO  [goThreadPoolTaskScheduler-6] ConsoleActivityMonitor:111 - Job 'tech-docs/131/opglive-europe-west4/1/deploy-docker' hung for more than '2857' minutes
2024-04-05 13:37:22,753 INFO  [goThreadPoolTaskScheduler-6] ConsoleActivityMonitor:111 - Job 'grafana-dashboards/382/sandbox-europe-west4/1/deploy-docker' hung for more than '1506' minutes
2024-04-05 13:37:22,753 INFO  [goThreadPoolTaskScheduler-6] ConsoleActivityMonitor:111 - Job 'grafana-dashboards/382/opglive-europe-west4/1/deploy-docker' hung for more than '1505' minutes
2024-04-05 13:37:22,753 INFO  [goThreadPoolTaskScheduler-6] ConsoleActivityMonitor:111 - Job 'grafana-dashboards/382/stage/1/deploy-docker' hung for more than '1507' minutes
2024-04-05 13:37:22,753 INFO  [goThreadPoolTaskScheduler-6] ConsoleActivityMonitor:111 - Job 'grafana-dashboards/382/stage/1/deploy' hung for more than '1507' minutes
2024-04-05 13:37:52,754 INFO  [goThreadPoolTaskScheduler-10] ConsoleActivityMonitor:111 - Job 'tech-docs/131/opglive-europe-west4/1/deploy-docker' hung for more than '2857' minutes
2024-04-05 13:37:52,754 INFO  [goThreadPoolTaskScheduler-10] ConsoleActivityMonitor:111 - Job 'grafana-dashboards/382/sandbox-europe-west4/1/deploy-docker' hung for more than '1506' minutes
2024-04-05 13:37:52,754 INFO  [goThreadPoolTaskScheduler-10] ConsoleActivityMonitor:111 - Job 'grafana-dashboards/382/opglive-europe-west4/1/deploy-docker' hung for more than '1505' minutes
2024-04-05 13:37:52,754 INFO  [goThreadPoolTaskScheduler-10] ConsoleActivityMonitor:111 - Job 'grafana-dashboards/382/stage/1/deploy-docker' hung for more than '1507' minutes
2024-04-05 13:37:52,754 INFO  [goThreadPoolTaskScheduler-10] ConsoleActivityMonitor:111 - Job 'grafana-dashboards/382/stage/1/deploy' hung for more than '1507' minutes
2024-04-05 13:38:13,382 INFO  [qtp380095890-24] BuildRepositoryRemoteImpl:96 - [Agent [opglive-europe-west1-gocd-agent-0, 10.100.36.3, 781d843e-1f98-4142-ac1e-80b2f6a94cf5, b78bd3e3-bc67-4df6-a3e5-b1d740507385]] is reporting result [Passed] for [Build [ppro-girogate/7/opglive-europe-west1/1/deploy-gcp/51500]]
2024-04-05 13:38:13,452 INFO  [qtp380095890-24] BuildRepositoryRemoteImpl:96 - [Agent [opglive-europe-west1-gocd-agent-0, 10.100.36.3, 781d843e-1f98-4142-ac1e-80b2f6a94cf5, b78bd3e3-bc67-4df6-a3e5-b1d740507385]] is reporting status [Completing] for [Build [ppro-girogate/7/opglive-europe-west1/1/deploy-gcp/51500]]
2024-04-05 13:38:13,696 INFO  [qtp380095890-35] BuildRepositoryRemoteImpl:96 - [Agent [opglive-europe-west1-gocd-agent-0, 10.100.36.3, 781d843e-1f98-4142-ac1e-80b2f6a94cf5, b78bd3e3-bc67-4df6-a3e5-b1d740507385]] is reporting status and result [Completed, Passed] for [Build [ppro-girogate/7/opglive-europe-west1/1/deploy-gcp/51500]]
2024-04-05 13:38:13,698 INFO  [qtp380095890-35] Stage:236 - Stage is being completed by transition id: 305134
```

- API call results before a restart:
```
{
  "name" : "opglive-europe-west1",
  "counter" : 1,
  "approval_type" : "manual",
  "approved_by" : "operations",
  "scheduled_at" : 1712317001126,
  "last_transitioned_time" : 1712317093697,
  "result" : "Passed",
  "rerun_of_counter" : null,
  "fetch_materials" : true,
  "clean_working_directory" : false,
  "artifacts_deleted" : false,
  "pipeline_name" : "ppro-girogate",
  "pipeline_counter" : 7,
  "jobs" : [ {
    "name" : "deploy-gcp",
    "state" : "Completed",
    "result" : "Passed",
    "scheduled_date" : 1712317001125,
    "rerun" : false,
    "original_job_id" : null,
    "agent_uuid" : "781d843e-1f98-4142-ac1e-80b2f6a94cf5",
    "pipeline_name" : null,
    "pipeline_counter" : null,
    "stage_name" : null,
    "stage_counter" : null,
    "job_state_transitions" : [ {
      "state" : "Scheduled",
      "state_change_time" : 1712317001125
    }, {
      "state" : "Assigned",
      "state_change_time" : 1712317006061
    }, {
      "state" : "Preparing",
      "state_change_time" : 1712317023480
    }, {
      "state" : "Building",
      "state_change_time" : 1712317029135
    }, {
      "state" : "Completing",
      "state_change_time" : 1712317093383
    }, {
      "state" : "Completed",
      "state_change_time" : 1712317093697
    } ]
  } ]
}#
```

- API call after the restart:

```
{
  "name" : "opglive-europe-west1",
  "counter" : 1,
  "approval_type" : "manual",
  "approved_by" : "operations",
  "scheduled_at" : 1712317001126,
  "last_transitioned_time" : 1712317093697,
  "result" : "Passed",
  "rerun_of_counter" : null,
  "fetch_materials" : true,
  "clean_working_directory" : false,
  "artifacts_deleted" : false,
  "pipeline_name" : "ppro-girogate",
  "pipeline_counter" : 7,
  "jobs" : [ {
    "name" : "deploy-gcp",
    "state" : "Completed",
    "result" : "Passed",
    "scheduled_date" : 1712317001125,
    "rerun" : false,
    "original_job_id" : null,
    "agent_uuid" : "781d843e-1f98-4142-ac1e-80b2f6a94cf5",
    "pipeline_name" : null,
    "pipeline_counter" : null,
    "stage_name" : null,
    "stage_counter" : null,
    "job_state_transitions" : [ {
      "state" : "Scheduled",
      "state_change_time" : 1712317001125
    }, {
      "state" : "Assigned",
      "state_change_time" : 1712317006061
    }, {
      "state" : "Preparing",
      "state_change_time" : 1712317023480
    }, {
      "state" : "Building",
      "state_change_time" : 1712317029135
    }, {
      "state" : "Completing",
      "state_change_time" : 1712317093383
    }, {
      "state" : "Completed",
      "state_change_time" : 1712317093697
    } ]
  } ]
}
```

--- a/api/api-stage-instance-v3/src/main/java/com/thoughtworks/go/apiv3/stageinstance/representers/StageRepresenter.java
+++ b/api/api-stage-instance-v3/src/main/java/com/thoughtworks/go/apiv3/stageinstance/representers/StageRepresenter.java
@@ -27,29 +27,30 @@ public class StageRepresenter {
         jsonWriter.add("approval_type", stage.getApprovalType());
         jsonWriter.add("approved_by", stage.getApprovedBy());
         jsonWriter.add("scheduled_at", stage.getCreatedTime().getTime());
-        jsonWriter.add("last_transitioned_time", stage.getLastTransitionedTime().getTime());
+        if (stage.getLastTransitionedTime() != null) {
+            jsonWriter.add("last_transitioned_time", stage.getLastTransitionedTime().getTime());
+        }
         if (stage.getResult() != null) {
             jsonWriter.add("result", stage.getResult().toString());
             if (stage.getResult() == StageResult.Cancelled) {
-                jsonWriter.add("cancelled_by", stage.getCancelledBy() == null? "GoCD" : stage.getCancelledBy());
+                jsonWriter.add("cancelled_by", stage.getCancelledBy() == null ? "GoCD" : stage.getCancelledBy());
             }
         }
         if (stage.getRerunOfCounter() == null) {
             jsonWriter.add("rerun_of_counter", (String) null);
-        }
-        else {
+        } else {
             jsonWriter.add("rerun_of_counter", stage.getRerunOfCounter());
         }
         jsonWriter.add("fetch_materials", stage.shouldFetchMaterials());
         jsonWriter.add("clean_working_directory", stage.shouldCleanWorkingDir());
         jsonWriter.add("artifacts_deleted", stage.isArtifactsDeleted());
         if (stage.getIdentifier() != null) {
-         jsonWriter.add("pipeline_name", stage.getIdentifier().getPipelineName());
-         jsonWriter.add("pipeline_counter", stage.getIdentifier().getPipelineCounter());
+            jsonWriter.add("pipeline_name", stage.getIdentifier().getPipelineName());
+            jsonWriter.add("pipeline_counter", stage.getIdentifier().getPipelineCounter());
         }
         jsonWriter.addChildList("jobs", jobsWriter -> stage.getJobInstances().forEach(
-                jobInstance -> jobsWriter.addChild(
-                        jobInstanceWriter -> JobInstanceRepresenter.toJSON(jobInstanceWriter, jobInstance))));
+            jobInstance -> jobsWriter.addChild(
+                jobInstanceWriter -> JobInstanceRepresenter.toJSON(jobInstanceWriter, jobInstance))));
 
     }
 }

